src/classes/RD2_ScheduleService.cls

```diff
@@ -364,6 +529,53 @@ public without sharing class RD2_ScheduleService {
         return schedulesByRdId;
     }
 
+    /***
+    * @description Retrieves recurring donation schedules associated with a single recurring donation
+    * @param rd npe03__Recurring_Donation__c
+    */
+    public List<RecurringDonationSchedule__c> getExistingSchedules(npe03__Recurring_Donation__c rd) {
+        if (rd == null) {
+            return null;
+        }
+
+        return getExistingSchedules(new List<npe03__Recurring_Donation__c> {rd}).get(rd.Id);
+    }
+
+    /***
+    * @description Retrieves recurring donation schedules associated with recurring donation list
+    * @param rds List<npe03__Recurring_Donation__c>
+    */
+    public Map<Id, List<RecurringDonationSchedule__c>> getExistingSchedules(List<npe03__Recurring_Donation__c> rds) {
```

Comment:
I have a suggestion for this. When the Schedule Service is called from the OpportunityEvaluationService class for the nightly job and queuable, the RD record will already have a sub-query on active RecurringDonationSchedule records (see [this method](https://github.com/SalesforceFoundation/NPSP/blob/feature/RD2/src/classes/RD2_OpportunityEvaluationService.cls#L254)). So, for this method it might be helpful if it first checked to see if there's an subquery on the RD (`RecurringDonationSchedules__r != null` possibly) to avoid an extra query when the visualization service is called from the OppEvalService. You can modify that query to add the foreign relationships if needed.

Reply:
I can do this, but doesn't that create a pretty tight linkage between these two pieces of code?  Let's discuss Monday.

Reply:
It doesn't have create a linkage. What I'm thinking is to use the sub-query if it's there, otherwise initiate a query. 

Maybe something like the following:
```
Set<Id> rdsToQuery = new Set<Id>();
for (npe03__Recurring_Donation__c rd : rds) {
    // might need to apply a namespace to the field in the keyset. Unsure.
    if (rd.getPopulatedFieldsAsMap().keySet().containsKey('RecurringDonationSchedules__r') {
        oldSchedules.put(rd.Id, rd.RecurringDonationSchedules__r);
    } else {
        rdsToQuery.add(rd.Id);
    }
}

if (rdsToQuery.isEmpty()) {
    return oldSchedules;
} 

// the rest of the logic to query and loop goes here
```

Reply:
I'd be happier with providing an override to the visualization method that allows the user to pass in a map of RD Ids, to schedules.  That makes it an explicit request on the part of the user to use the passed in schedules to determine the installments.  The aspect of your suggestion that makes me uncomfortable is the implicit rather than explicit nature of the request.